Rubberduck.VBEEditor/ComManagement/TypeLibs/TypeLibs.cs

```diff
@@ -0,0 +1,1669 @@
+    /// Encapsulates reading unmanaged memory into managed structures
+    /// </summary>
+    public static class StructHelper
+    {
+        /// <summary>
+        /// Takes a COM object, and reads the unmanaged memory given by its pointer, allowing us to read internal fields
+        /// </summary>
+        /// <typeparam name="T">the type of structure to return</typeparam>
+        /// <param name="comObj">the COM object</param>
+        /// <returns>the requested structure T</returns>
+        public static T ReadComObjectStructure<T>(object comObj)
+        {
+            // Reads a COM object as a structure to copy its internal fields
+            if (Marshal.IsComObject(comObj))
+            {
+                var referencesPtr = Marshal.GetIUnknownForObject(comObj);
+                var retVal = StructHelper.ReadStructure<T>(referencesPtr);
+                Marshal.Release(referencesPtr);
+                return retVal;
+            }
+            else
+            {
+                throw new ArgumentException("Expected a COM object");
+            }
+        }
+
+        /// <summary>
+        /// Takes an unmanaged memory address and reads the unmanaged memory given by its pointer
+        /// </summary>
+        /// <typeparam name="T">the type of structure to return</typeparam>
+        /// <param name="memAddress">the unamanaged memory address to read</param>
+        /// <returns>the requested structure T</returns>
+        public static T ReadStructure<T>(IntPtr memAddress)
+        {
+            if (memAddress == IntPtr.Zero) return default(T);
+            return (T)Marshal.PtrToStructure(memAddress, typeof(T));
+        }
+
+        /// <summary>
+        /// Takes an unmanaged memory address and reads the unmanaged memory given by its pointer, 
+        /// with memory address validation for added protection
+        /// </summary>
+        /// <typeparam name="T">the type of structure to return</typeparam>
+        /// <param name="memAddress">the unamanaged memory address to read</param>
+        /// <returns>the requested structure T</returns>
+        public static T ReadStructureSafe<T>(IntPtr memAddress)
+        {
+            if (memAddress == IntPtr.Zero) return default(T);
+            return (T)Marshal.PtrToStructure(memAddress, typeof(T));
+        }
+    }
+
+    /// <summary>
+    /// Ensures that a wrapped COM object only responds to a specific COM interface.
+    /// </summary>
+    /// <typeparam name="T">The COM interface for restriction</typeparam>
+    public class RestrictComInterfaceByAggregation<T> : ICustomQueryInterface, IDisposable
+    {
+        private IntPtr _outerObject;
+        private T _wrappedObject;
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="outerObject">The object that needs interface requests filtered</param>
+        /// <param name="queryForType">determines whether we call QueryInterface for the interface or not</param>
```

Comment:
Also, I think it's good if you document the [`returns`](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/returns), so that the possibility of the `null` is called out in the XML-docs itself.

Reply:
sure, will make this clearer.  there is a good reason on this one